<?xml version="1.0"?>
<launch>

    <arg name="r_name" default="racecar"/>
    <arg name="vesc_ns" default="/vesc"/>

    <!--Launch the move_base node-->
    <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen" ns="/$(arg r_name)">

        <remap from="/$(arg r_name)/$(arg r_name)/scan" to="/$(arg r_name)/scan"/>
        <remap from="/$(arg r_name)/odom" to="$(arg vesc_ns)/odom"/>

        <rosparam file="$(find racecar_navigation)/params/costmap_common_params.yaml" command="load" ns="global_costmap"/>
        <rosparam param="/$(arg r_name)/move_base/global_costmap/obstacle_layer/laser_scan_sensor/topic" subst_value="True">$(arg r_name)/scan</rosparam>
        <rosparam param="/$(arg r_name)/move_base/global_costmap/obstacle_layer/laser_scan_sensor/frame_name" subst_value="True">$(arg r_name)/laser</rosparam>

        <rosparam file="$(find racecar_navigation)/params/costmap_common_params.yaml" command="load" ns="local_costmap"/>
        <rosparam param="/$(arg r_name)/move_base/local_costmap/obstacle_layer/laser_scan_sensor/topic" subst_value="True">$(arg r_name)/scan</rosparam>
        <rosparam param="/$(arg r_name)/move_base/local_costmap/obstacle_layer/laser_scan_sensor/frame_name" subst_value="True">$(arg r_name)/laser</rosparam>

        <rosparam file="$(find racecar_navigation)/params/local_costmap_params.yaml" command="load"/>
        <rosparam param="/$(arg r_name)/move_base/local_costmap/robot_base_frame" subst_value="True">$(arg r_name)/base_link</rosparam>

        <rosparam file="$(find racecar_navigation)/params/global_costmap_params.yaml" command="load"/>
        <rosparam param="/$(arg r_name)/move_base/global_costmap/robot_base_frame" subst_value="True">$(arg r_name)/base_link</rosparam>

        <rosparam file="$(find racecar_navigation)/params/teb_planner_params.yaml" command="load"/>
        <rosparam param="/$(arg r_name)/move_base/TebLocalPlannerROS/odom_topic" subst_value="True">$(arg vesc_ns)/odom</rosparam>
    </node>

    <node pkg="racecar_navigation" type="threeLanes_current_future_pos.py" name="threeLanes_current_future_pos" output="screen" ns="/$(arg r_name)">
    <remap from="/vesc/odom" to="$(arg vesc_ns)/odom"/>
    </node>
    
    <node pkg="racecar_navigation" type="chosen_nav_action.py" name="chosen_nav_action" output="screen" ns="/$(arg r_name)"/>
    <node pkg="racecar_navigation" type="isGoalReached.py" name="isGoalReached" output="screen" ns="/$(arg r_name)">
    <remap from="/vesc/odom" to="$(arg vesc_ns)/odom"/>
    </node>
    
    <!--<node pkg="racecar_navigation" type="action_client_LaneKeeping.py" name="action_client_LaneKeeping" output="screen" ns="/$(arg r_name)">-->

    <!--<node pkg="racecar_navigation" type="action_server_LaneKeeping.py" name="action_server_LaneKeeping" output="screen" ns="/$(arg r_name)">
    <remap from="/vesc/odom" to="$(arg vesc_ns)/odom"/>
    <remap from="/camera/image_raw" to="/$(arg r_name)/camera/image_raw"/>
    <remap from="/drive_parameters" to="/$(arg r_name)/drive_parameters"/>
    </node>-->
    
    <node pkg="racecar_move_car" type="move_car_client.py" name="move_car_client" output="screen" ns="/$(arg r_name)">
    </node>

    <!-- drive_param to servo node -->
    <node pkg="racecar_control" type="drive_param_to_servo.py" name="drive_param_to_servo" output="screen" ns="/$(arg r_name)">
    <remap from="/vesc/ackermann_cmd_mux/input/teleop" to="$(arg vesc_ns)/ackermann_cmd_mux/input/teleop"/>
    <remap from="/drive_parameters" to="/$(arg r_name)/drive_parameters"/>
    </node>

</launch>
